### PR TITLE
CP-13493: Thread to update xapi db for physical utilisation of SR

### PR DESCRIPTION
**Pull request by sharady**
This new thread will update physical utilisation only for SR
which has SR_STATS capability

Signed-off-by: Sharad Yadav <sharad.yadav@citrix.com>

### CODE REVIEW

File: ocaml/xapi/xapi_sr.ml [hunk 1]
```diff
@@ -128,6 +130,28 @@ let scanning_thread () = Debug.with_thread_named "scanning_thread" (fun () ->
...
+	let srs = get_all_plugged_srs ~__context in
+	(* Filter the SRs which have SR_STATS capability *)
+	let rrd_srs = List.filter (fun sr ->
+		let sr_record = Db.SR.get_record_internal ~__context ~self:sr in
+		Smint.(has_capability Sr_stats (Xapi_sr_operations.features_of_sr ~__context sr_record))
+		) srs
+	in
+	(* Update the physical utilisation db field of SRs *)
+	List.iter (fun sr ->
+		let new_value = Rrdd.query_sr_ds ~sr_uuid:(Db.SR.get_uuid ~__context ~self:sr) ~ds_name:"physical_utilisation" in
```
Comment (simonjbeaumont) [inline ocaml/xapi/xapi_sr.ml]: The design calls for some form of caching to save us from doing this work every time: http://xapi-project.github.io/rrdd/futures/sr-level-rrds.html

**Comment by simonjbeaumont** [comment]
This isn't quite what the "Database updating" section of the design says at http://xapi-project.github.io/rrdd/futures/sr-level-rrds.html. It says that it should probably run on every host and only update the SRs for which it is the SRmaster. This is subtley different from only running the thread itself on the master.

File: ocaml/xapi/xapi_sr.ml [hunk 2]
```diff
@@ -128,6 +130,28 @@ let scanning_thread () = Debug.with_thread_named "scanning_thread" (fun () ->
 		done)
 	) ()
 
+let update_physical_utilisation ~__context =
+	let srs = get_all_plugged_srs ~__context in
+	(* Filter the SRs which have SR_STATS capability *)
+	let rrd_srs = List.filter (fun sr ->
+		let sr_record = Db.SR.get_record_internal ~__context ~self:sr in
```
Comment (simonjbeaumont) [inline ocaml/xapi/xapi_sr.ml]: Do we need the _internal version here? We might do, just checking!

**Comment by simonjbeaumont** [comment]
Build failed:=
```
08:00:10 File "xapi_sr.ml", line 19, characters 0-31:
08:00:10 Error: Multiple definition of the module name Rrdd.
08:00:10        Names must be unique in a given structure or signature.
```

**Comment by sharady** [comment]
looking into that.

**Comment by sharady** [comment]
need to rebase, as per the changes gone in https://github.com/xapi-project/xen-api/commit/1345299e9117120087119a111d8f6f7294ee9c67

File: ocaml/xapi/xapi_sr.ml [hunk 3]
```diff
@@ -124,6 +124,34 @@ let scanning_thread () = Debug.with_thread_named "scanning_thread" (fun () ->
 		done)
 	) ()
 
+let sr_cache = ref []
+
+let update_physical_utilisation ~__context =
+	let srs = Helpers.get_all_plugged_srs ~__context in
+	(* Remove SRs from sr_cache if they are no more available *)
+	let remove_srs = List.filter (fun sr -> (List.mem sr srs) = false ) !sr_cache in
+	sr_cache := List.filter (fun sr -> (List.mem sr remove_srs) = false ) !sr_cache;
```
Comment (simonjbeaumont) [inline ocaml/xapi/xapi_sr.ml]: I still think these two lines could be combined more neatly:
```
(* Remove items from the cache that are not in srs *)
sr_cache := List.filter (fun sr -> List.mem sr srs) !sr_cache;
```

File: ocaml/xapi/xapi_sr.ml [hunk 4]
```diff
@@ -124,6 +124,34 @@ let scanning_thread () = Debug.with_thread_named "scanning_thread" (fun () ->
...
+	(* Checks whether SR is sr_master and has SR_STATS capability *)
+	let is_stats_sr ~__context sr =
+		let sr_record = Db.SR.get_record_internal ~__context ~self:sr in
+		if (Smint.(has_capability Sr_stats (Xapi_sr_operations.features_of_sr ~__context sr_record)) &&
+			(Helpers.i_am_srmaster ~__context ~sr))
+			then true else false
+		in
```
Comment (simonjbeaumont) [inline ocaml/xapi/xapi_sr.ml]: This `in` should either be on the end of the previous line or aligned with the `let`.

**Comment by simonjbeaumont** [comment]
This looks good. Can you now squash these commits and we'll merge to trunk-rign3 and open a duplicate pull request (referencing this one) to trunk-thin-lvhd-2

**Comment by simonjbeaumont** [comment]
Superseded by #2388 for feature branch.

**Pull request closed without merging**